firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionTesterApiClient.java

```diff
@@ -109,62 +96,116 @@ Task<AppDistributionReleaseInternal> fetchNewRelease() {
-    String url = String.format("https://%s/%s", APP_TESTERS_HOST, path);
-    HttpsURLConnection connection = null;
-    try {
-      connection = openHttpsUrlConnection(url, token);
-      return readResponse(connection);
-    } catch (IOException e) {
-      throw new FirebaseAppDistributionException(
-          ErrorMessages.NETWORK_ERROR, Status.NETWORK_FAILURE, e);
-    } finally {
-      if (connection != null) {
-        connection.disconnect();
-      }
-    }
+  /**
+   * Fetches and returns the name of the installed release given the IAS artifact ID of the
+   * installed app bundle, or null if it could not be found.
+   */
+  @NonNull
+  Task<String> findReleaseUsingIasArtifactId(String iasArtifactId) {
+    return runWithFidAndToken(
+        (fid, token) -> findRelease(fid, token, FIND_RELEASE_IAS_ARTIFACT_ID_PARAM, iasArtifactId));
   }
 
-  private String readResponseBody(HttpsURLConnection connection) throws IOException {
-    boolean isSuccess = isResponseSuccess(connection.getResponseCode());
-    try (InputStream inputStream =
-        isSuccess ? connection.getInputStream() : connection.getErrorStream()) {
-      if (inputStream == null && !isSuccess) {
-        // If the server returns a response with an error code and no response body, getErrorStream
-        // returns null. We return an empty string to reflect the empty body.
-        return "";
-      }
-      return convertInputStreamToString(new BufferedInputStream(inputStream));
+  private String findRelease(String fid, String token, String binaryIdParam, String binaryIdValue)
+      throws FirebaseAppDistributionException {
+    String path =
+        String.format(
+            "v1alpha/projects/%s/installations/%s/releases:find?%s=%s",
+            firebaseApp.getOptions().getGcmSenderId(), // Project number
+            fid,
+            binaryIdParam,
+            binaryIdValue);
+    HttpResponse response = testerApiHttpClient.makeGetRequest(path, token);
+    if (response.code() == 404) {
+      return null;
+    } else if (!response.isSuccess()) {
+      throw getExceptionForHttpResponse(response.code());
     }
+    return parseJsonFieldFromResponse(response.body(), "release");
   }
 
-  private static boolean isResponseSuccess(int responseCode) {
-    return responseCode >= 200 && responseCode < 300;
+  /** Creates a new feedback from the given text, and returns the feedback name. */
+  @NonNull
+  Task<String> createFeedback(String testerReleaseName, String feedbackText) {
+    return runWithFidAndToken(
+        (unused, token) -> {
+          LogWrapper.getInstance().i("Creating feedback for release: " + testerReleaseName);
+          String path = String.format("v1alpha/%s/feedback", testerReleaseName);
+          String requestBody = buildCreateFeedbackBody(feedbackText).toString();
+          HttpResponse response = testerApiHttpClient.makePostRequest(path, token, requestBody);
+          if (response.code() == 403) {
```

Comment:
why handle this separately? I think `getExceptionForHttpResponse()` would do the same